artichoke-backend/src/extn/core/time/args.rs

```diff
@@ -21,134 +21,133 @@ impl Default for Args {
             hour: 0,
             minute: 0,
             second: 0,
-            micros: 0,
+            nanoseconds: 0,
         }
     }
 }
 
-impl Args {
-    pub fn year(&self) -> Result<i32, Error> {
-        i32::try_from(self.year).map_err(|_| ArgumentError::with_message("year out of range").into())
-    }
+impl TryConvertMut<&[Value], Args> for Artichoke {
+    type Error = Error;
 
-    pub fn month(&self) -> Result<u8, Error> {
-        match u8::try_from(self.month) {
-            Ok(month @ 1..=12) => Ok(month),
-            _ => Err(ArgumentError::with_message("mon out of range").into()),
+    fn try_convert_mut(&mut self, args: &[Value]) -> Result<Args, Self::Error> {
+        // Time args should have a length of 1..=8 or 10. The error does not
+        // give a hint that the 10 arg variant is supported however (this is
+        // the same in MRI).
+        if let 0 | 9 | 11 = args.len() {
+            let mut message = br#"wrong number of arguments (given "#.to_vec();
+            message.extend(args.len().to_string().bytes());
+            message.extend_from_slice(b", expected 1..8)");
+            return Err(ArgumentError::from(message).into());
         }
-    }
 
-    pub fn day(&self) -> Result<u8, Error> {
-        match u8::try_from(self.day) {
-            Ok(day @ 1..=31) => Ok(day),
-            _ => Err(ArgumentError::with_message("mday out of range").into()),
-        }
-    }
+        let mut args = args.to_vec();
 
-    pub fn hour(&self) -> Result<u8, Error> {
-        match u8::try_from(self.hour) {
-            Ok(hour @ 0..=23) => Ok(hour),
-            _ => Err(ArgumentError::with_message("hour out of range").into()),
+        // Args are in order of year, month, day, hour, minute, second, micros.
+        // This is unless there are 10 arguments provided (`Time#to_a` format),
+        // at which points it is second, minute, hour, day, month, year.
+        if args.len() == 10 {
+            args.swap(0, 5);
+            args.swap(1, 4);
+            args.swap(2, 3);
+            // All arguments after position 5 are ignored in the 10 argument
+            // variant.
+            args.truncate(6);
         }
-    }
 
-    pub fn minute(&self) -> Result<u8, Error> {
-        match u8::try_from(self.minute) {
-            Ok(minute @ 0..=59) => Ok(minute),
-            _ => Err(ArgumentError::with_message("min out of range").into()),
-        }
-    }
+        let mut result = Args::default();
 
-    pub fn second(&self) -> Result<u8, Error> {
-        match u8::try_from(self.second) {
-            Ok(second @ 0..=60) => Ok(second),
-            _ => Err(ArgumentError::with_message("sec out of range").into()),
-        }
-    }
+        for (i, &arg) in args.iter().enumerate() {
+            match i {
+                0 => {
+                    let arg = to_int(self, arg)?;
+                    let arg: i64 = arg.try_convert_into::<Option<i64>>(self)?.unwrap();
```

Comment:
I'm a bit confused why we convert to `Option<i64>` and then unwrap. Can we replace this with this:

```suggestion
                    let arg = arg.try_convert_into::<i64>(self)?;
```

Reply:
You know what, I'm also confused as to why I put it there in the original code too :grimacing:. I assume `to_int` should never result in a `nil`, so easy fix!